clippy_lints/src/non_send_field_in_send_ty.rs

```diff
@@ -0,0 +1,229 @@
+
+impl<'tcx> LateLintPass<'tcx> for NonSendFieldInSendTy {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'_>) {
+        let ty_allowed_in_send = if self.enable_raw_pointer_heuristic {
+            ty_allowed_with_raw_pointer_heuristic
+        } else {
+            ty_implements_send_or_copy
+        };
+
+        // Checks if we are in `Send` impl item.
+        // We start from `Send` impl instead of `check_field_def()` because
+        // single `AdtDef` may have multiple `Send` impls due to generic
+        // parameters, and the lint is much easier to implement in this way.
+        if_chain! {
+            if let Some(send_trait) = cx.tcx.get_diagnostic_item(sym::send_trait);
+            if let ItemKind::Impl(hir_impl) = &item.kind;
+            if let Some(trait_ref) = &hir_impl.of_trait;
+            if let Some(trait_id) = trait_ref.trait_def_id();
+            if send_trait == trait_id;
+            if let ImplPolarity::Positive = hir_impl.polarity;
+            if let Some(ty_trait_ref) = cx.tcx.impl_trait_ref(item.def_id);
+            if let self_ty = ty_trait_ref.self_ty();
+            if let ty::Adt(adt_def, impl_trait_substs) = self_ty.kind();
+            then {
+                let mut non_send_fields = Vec::new();
+
+                let hir_map = cx.tcx.hir();
+                for variant in &adt_def.variants {
+                    for field in &variant.fields {
+                        if_chain! {
+                            if let Some(field_hir_id) = field
+                                .did
+                                .as_local()
+                                .map(|local_def_id| hir_map.local_def_id_to_hir_id(local_def_id));
+                            if !is_lint_allowed(cx, NON_SEND_FIELD_IN_SEND_TY, field_hir_id);
+                            if let field_ty = field.ty(cx.tcx, impl_trait_substs);
+                            if !ty_allowed_in_send(cx, field_ty, send_trait);
+                            if let Some(field_span) = hir_map.span_if_local(field.did);
+                            then {
+                                non_send_fields.push(NonSendField {
+                                    name: hir_map.name(field_hir_id),
+                                    span: field_span,
+                                    ty: field_ty,
+                                    generic_params: collect_generic_params(cx, field_ty),
+                                })
+                            }
+                        }
+                    }
+                }
+
+                if !non_send_fields.is_empty() {
+                    span_lint_and_then(
+                        cx,
+                        NON_SEND_FIELD_IN_SEND_TY,
+                        item.span,
+                        &format!(
+                            "this implementation is unsound, as some fields in `{}` are `!Send`",
+                            self_ty
```

Comment:
The display function of `Ty<>` sadly prefixed the entire path to the type, this can result in lint message with long paths. To keep the message a bit more concise, it would be better if the type can be copied from the source code:
```rs
clippy_utils::source::snippet(cx, self_ty.span, "..")
```

See #7212 for further information. The same thing should be applied for the help and note messages :upside_down_face:

Reply:
Using the code span will rarely show incomplete type. For instance, consider the output of `Complex` testcase.

```
error: this implementation is unsound, as some fields in `Complex<Q, std::sync::MutexGuard<'static, bool>>` are `!Send`
  --> $DIR/non_send_field_in_send_ty.rs:125:1
   |
LL | unsafe impl<Q: Send> Send for Complex<Q, MutexGuard<'static, bool>> {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: the field `field2` has type `std::sync::MutexGuard<'static, bool>` which is `!Send`
  --> $DIR/non_send_field_in_send_ty.rs:119:5
   |
LL |     field2: B,
   |     ^^^^^^^^^
   = help: use a thread-safe type that implements `Send`
```

If we use span, the output message would become "has type `B`" although `B` is replaced with a concrete type `MutexGuard<'static, bool>`. Would this be okay?

Reply:
Decided to omit the type in the "note" part, as it becomes redundant with the code snippet if we use span to print it.